### PR TITLE
Implement prefix and range scans, return enumerators from `each`_` without block, extension code cleanup

### PR DESCRIPTION
**Pull request by y8**
* RocksDB can be fully configured by Option String
* Get methods now `get_one` and `get_many`. `get` method can automaticaly use appropriate method depending on the number of arguments
* User friendly enumerable names: `each`, `reverse_each`, `each_key`, `each_pair`
* Added `each_prefix`,  `each_range` methods to efficenly scan by prefix and value ranges
* Added `previous` and `seek_for_previous` methods to iterator
* Extended spec coverage, with multiple corner cases
* Slightly DRY-ed extension code, with additional safety checks
* Updated README with more examples

---

When upgrading from `0.2` version, please note the following breaking changes:

* `multi_get` will return `nil` instead of empty string
* `RocksDB::DB.get_instance` is removed. Implement your own DB instance cache if needed
* `each_` methods now returns `Enumerator` instead of `RocksDB::Iterator`

Also, there some things that are now deprecated and will be removed in future versions:

* `RocksDB::DBError` was deprecated and replaced with `RocksDB::Error`. Specific errors now have their own exception class: `ReadOnly`, `DatabaseClosed`, `IteratorClosed`, `StatusError`
* `is_open?` was replaced with `open?`
* `is_readonly?` was replaced with `writable?`
* `multi_get` was replaced with `get_many`
* `new_iterator` was replaced with `to_iterator`
* `Iterator#valid` was replaced with `Iterator#valid?`

### CODE REVIEW

File: lib/rocksdb.rb [hunk 1]
```diff
@@ -1,61 +1,72 @@
...
+        get_many args
       end
     end
+
+    alias_method :includes?, :exists?
+    alias_method :contains?, :exists?
+    alias_method :member?, :exists?
+    alias_method :member?, :exists?
+    alias_method :[], :get
+    alias_method :[]=, :put
```
Comment (katafrakt) [inline lib/rocksdb.rb]: I'm not 100% sure about this. I understand it's Ruby-ish and cool, but in general I'm not a fan of having multiple ways to do exactly the same thing.

On related note, `database['something']` is not a clear concept at the first glance - is it fetching data or maybe some configuration option? `database.get` is rather clear in that regard.

**Comment by katafrakt** [comment]
Cool batch of changes! I agree with most of them, however I have some doubts regarding adding many aliases to existing methods. Anyway, great job in my opinion,

File: lib/rocksdb.rb (hunk 1 above)
Comment (y8) [reply lib/rocksdb.rb]: I kept this part of interface as it was: 

https://github.com/isamu/rocksdb-ruby/blob/abe4be82d6d2291f10a30c426af66a9f8fa55326/lib/rocksdb.rb#L47-L48

It's a quite common thing in the ruby world to have multiple names for the same thing to make code more readable. 

I understand your concern about `[]`, and there is no consistency between other gems. Few popular ones ([moneta-rb](https://github.com/moneta-rb/moneta), [daybreak](https://propublica.github.io/daybreak/), [riak](https://github.com/basho/riak-ruby-client))  have `[]` and `[]=` as well as `get` and `set` (I think it worth adding `set` along with `put`)

I don't have strong opinion about that, and will deprecate them if needed.

**Comment by y8** [comment]
Thanks! 

But I'm having hard time trying to find out that causes the segfault on travis.

I'm a total noob with c/cpp development. There is no references to the library in the c stacktrace and gdb `thread apply all bt full"` on a coredump shows literally nothing. 

https://travis-ci.org/y8/rocksdb-ruby/builds/603608782#L893
https://travis-ci.org/y8/rocksdb-ruby/builds/603608782#L1494

I've built rocksdb with `DEBUG_LEVEL=1` and added following to extconf:

```ruby
  # Make with debug flags
  if have_library('rocksdb_debug') or ENV.key?("DEBUG_LEVEL")
    warn "Building with debug symbols and without optimizations"
    RbConfig::CONFIG["optflags"] = "-O0"
    RbConfig::CONFIG["debugflags"] = "-ggdb3"
  end
```

What's my options here?

File: lib/rocksdb.rb (hunk 1 above)
Comment (katafrakt) [reply lib/rocksdb.rb]: > It's a quite common thing in the ruby world to have multiple names for the same thing to make code more readable.

Yes, I'm aware of that, but I still think that smaller API is a better API 😉 Then, similar to you, it's not a very strong opinion. Perhaps we need more people to express their opinion.

**Comment by katafrakt** [comment]
Perhaps something with RocksDB versions? I test locally with 3.1.2 (version available in Homebrew), but on Travis it runs against 3.6.3. I don't have this version at hand, maybe I'll check it on some Linux machine.

Which version are you testing locally with?

**Comment by y8** [comment]
I'm testing on macOS on 3.1.2 and 3.6.3 (I'll PR updated homebrew formula with zstd later today)

It was my mistake: I haven't provided ruby executable to gdb when openinn core dump. Now it works. 

Segfault is indeed a bug (why only on linux?) in `SLICE_FROM_RB_VALUE` and `STRING_FROM_RB_VALUE` macros. I'm passing pointer from `StringValuePtr`, but `RSTRING_LEN` is called on original string. 

Now I'm a bit stuck with that. Converting macro to function

```cpp
  rocksdb::Slice rocksdb_db_slice_from_value(VALUE entity) {
    VALUE string_value = StringValue(entity);

    return rocksdb::Slice(RSTRING_PTR(string_value), RSTRING_LEN(string_value));
  };
```

 doesn't work as function return non-C types `rocksdb::Slice` and `std::string`:

```
rocksdb_rb.cc:96:15: warning: 'rocksdb_slice_from_rb_value' has C-linkage specified, but returns user-defined type 'std::string' (aka
```

What is a standard way of dealing with that? Maybe there is a way to use temporary variable inside the macro expression to avoid calling functions?

**Comment by y8** [comment]
By the way, could you please enable Travis CI Pull Request Check?

cc @isamu

**Comment by isamu** [comment]
@y8  Thank you for your good patch.

You can check Travis CI result in  https://travis-ci.org/isamu/rocksdb-ruby .

**Comment by y8** [comment]
You can additionally enable Travis CI <-> Github integration. This way passing travis build will be added as check for a pull request

Like this:


Read more here: https://blog.travis-ci.com/2018-05-07-announcing-support-for-github-checks-api-on-travis-ci-com

**Comment by y8** [comment]
Awesome, thanks!

**Comment by y8** [comment]
Build passed. Ready for review

**Comment by y8** [comment]
And few more:

* Allow building on Linux and macOS, support building with a specific rocksdb version
* Build master on Linux, macOS with ruby 2.4, 2.5, 2.6 by default
* Build some specific versions on Linux and macOS
* Add `RocksDB.library_version` to return version of librocksdb
* Re-implement option string parsing for `options` method in ruby, as StringToMap is only available in 5.8 and later
* Disable `seek_for_previous` on 4.11 and earlier, as it not implemented

Now that's really it!

**Comment by isamu** [comment]
Thank you for your great changes.
I will release new version.

**Pull request merged**